django/db/models/expressions.py

```diff
@@ -843,13 +843,19 @@ def as_sql(self, compiler, connection, template=None, case_joiner=None, **extra_
         template_params.update(extra_context)
         case_parts = []
         sql_params = []
+        from django.db.models.sql.datastructures import EmptyResultSet
```

Comment:
Since EmptyResultSet isn't necessarily related to SQL I think moving it makes sense. `django.core.exceptions` is probably a good place to move it to. The only question is whether we need to deprecate the old location, or at least provide an import in the old location.

@akaariai @loic @manfre @adamchainz do any of you have an opinion here? I imagine non-core backends are likely to import this type but I haven't checked specifically.

In any case, if the change is made, it should be in a separate commit to catch regressions in the future without impacting 26517 if it needs to be reverted.